README.md
<h1 align="center">vue-grid-layout</h1>

<p align="center">
<a href="https://www.npmjs.com/package/vue-grid-layout"><img src="https://img.shields.io/npm/v/vue-grid-layout.svg"/> <img src="https://img.shields.io/npm/dm/vue-grid-layout.svg"/></a> <a href="https://vuejs.org/"><img src="https://img.shields.io/badge/vue-2.2.x-brightgreen.svg"/></a>
</p>

vue-grid-layout is a grid layout system, like [Gridster](http://dsmorse.github.io/gridster.js/), for Vue.js. **Heavily inspired by [React-Grid-Layout](https://github.com/STRML/react-grid-layout)**

### **Current version:** 2.3.9 (Supports Vue 2.2+)

### **For Vue 2.1.10 and below use version [2.1.3](https://github.com/jbaysolutions/vue-grid-layout/tree/2.1.3)**
### **For Vue 1 use version [1.0.3](https://github.com/jbaysolutions/vue-grid-layout/tree/1.0.3)** 

<br/>

[**[Demo](https://jbaysolutions.github.io/vue-grid-layout/examples/01-basic.html) | [Changelog](/CHANGELOG.md)**]

English | [简体中文](./README-zh_CN.md) 

<!--
## Table of Contents

- [Features](#features)
- [Getting Started](#getting-started)
  - [Installation](#installation)
    - [npm](#npm)
- [Contribute](#contribute)
- [TODO List](#todo-list)

## Demos


TODO UPDATE DOCS

Used guide for vue cli build: https://medium.com/justfrontendthings/how-to-create-and-publish-your-own-vuejs-component-library-on-npm-using-vue-cli-28e60943eed3
Also check https://cli.vuejs.org/guide/build-targets.html#library

-->

#### Projects using vue-grid-layout

- [Draxed](https://www.draxed.com/?utm_source=github&utm_medium=web&utm_campaign=vue-grid-layout)
- [cryptotiles](https://www.cryptotiles.io/?utm_source=github&utm_medium=web&utm_campaign=vue-grid-layout)
- [Data Providers](https://www.dataproviders.io/?utm_source=github&utm_medium=web&utm_campaign=vue-grid-layout)
- [Cataholic](https://cataholic.glitch.me/)

*Know of others? Create a PR to let me know!*

## Features

* Draggable widgets
* Resizable widgets
* Static widgets
* Bounds checking for dragging and resizing
* Widgets may be added or removed without rebuilding grid
* Layout can be serialized and restored
* Automatic RTL support (resizing not working with RTL on 2.2.0)
* Responsive


## Getting Started

### Installation

#### npm

    # install with npm    
	npm install vue-grid-layout --save
    
    # install with yarn    
    yarn add vue-grid-layout


Import the library

```javascript
    import VueGridLayout from 'vue-grid-layout';
```

Add to other Vue components 

 ```javascript
    export default {
        components: {
            GridLayout: VueGridLayout.GridLayout,
            GridItem: VueGridLayout.GridItem
        },
    // ... data, methods, mounted (), etc.
    }
    
```    

#### browser

Include the browser-ready bundle (download from [releases](https://github.com/jbaysolutions/vue-grid-layout/releases)) in your page. The components will be automatically available.

```html
    <script src="vue-grid-layout.umd.min.js"></script>
```

### Usage

```javascript
	var testLayout = [
	    {"x":0,"y":0,"w":2,"h":2,"i":"0"},
	    {"x":2,"y":0,"w":2,"h":4,"i":"1"},
	    {"x":4,"y":0,"w":2,"h":5,"i":"2"},
	    {"x":6,"y":0,"w":2,"h":3,"i":"3"},
	    {"x":8,"y":0,"w":2,"h":3,"i":"4"},
	    {"x":10,"y":0,"w":2,"h":3,"i":"5"},
	    {"x":0,"y":5,"w":2,"h":5,"i":"6"},
	    {"x":2,"y":5,"w":2,"h":5,"i":"7"},
	    {"x":4,"y":5,"w":2,"h":5,"i":"8"},
	    {"x":6,"y":3,"w":2,"h":4,"i":"9"},
	    {"x":8,"y":4,"w":2,"h":4,"i":"10"},
	    {"x":10,"y":4,"w":2,"h":4,"i":"11"},
	    {"x":0,"y":10,"w":2,"h":5,"i":"12"},
	    {"x":2,"y":10,"w":2,"h":5,"i":"13"},
	    {"x":4,"y":8,"w":2,"h":4,"i":"14"},
	    {"x":6,"y":8,"w":2,"h":4,"i":"15"},
	    {"x":8,"y":10,"w":2,"h":5,"i":"16"},
	    {"x":10,"y":4,"w":2,"h":2,"i":"17"},
	    {"x":0,"y":9,"w":2,"h":3,"i":"18"},
	    {"x":2,"y":6,"w":2,"h":2,"i":"19"}
	];
	
	new Vue({
	    el: '#app',
	    data: {
	        layout: testLayout,
	    },
	});
``` 


```html

    <grid-layout
            :layout.sync="layout"
            :col-num="12"
            :row-height="30"
            :is-draggable="true"
            :is-resizable="true"
            :is-mirrored="false"
            :vertical-compact="true"
            :margin="[10, 10]"
            :use-css-transforms="true"
    >

        <grid-item v-for="item in layout"
                   :x="item.x"
                   :y="item.y"
                   :w="item.w"
                   :h="item.h"
                   :i="item.i"
                   :key="item.i">
            {{item.i}}
        </grid-item>
    </grid-layout>
```


### Documentation

#### Properties

##### GridLayout

* **layout**
    
    * type: `Array`
    * required: `true`

    This is the initial layout of the grid.

    The value must be an `Array` of `Object` items. Each item must have `i`, `x`, `y`, `w` and `h` properties. Please refer to the documentation for `GridItem` below for more information.

* **responsiveLayouts**

    * type: `Object`
    * required: `false`
    * default : `{}`

    This is the initial layouts of the grid per breakpoint if `responsive` is set to `true`.
    The keys of the `Object` are breakpoint names and each value is an `Array` of `Object` items as defined by `layout` prop. eg:{ lg:[layout items], md:[layout items] }.
    Setting the prop after the creation of the GridLayout has no effect.

* **colNum**
    
    * type: `Number`
    * required: `false`
    * default: `12`

    Says how many columns the grid has.

    The value should be a _natural number_. 

* **rowHeight**
    
    * type: `Number`
    * required: `false`
    * default: `150`

    Says what is a height of a single row in pixels.

* **maxRows**
    
    * type: `Number`
    * required: `false`
    * default: `Infinity`

    Says what is a maximal number of rows in the grid.

* **margin**
    
    * type: `Array`
    * required: `false`
    * default: `[10, 10]`

    Says what are the margins of elements inside the grid.

    The value must be a two-element `Array` of `Number`. Each value is expressed in pixels. The first element is a margin horizontally, the second element is a vertical margin.

* **isDraggable**
    
    * type: `Boolean`
    * required: `false`
    * default: `true`

    Says if the grids items are draggable.

* **isResizable**
    
    * type: `Boolean`
    * required: `false`
    * default: `true`

    Says if the grids items are resizable.

* **isMirrored**
    
    * type: `Boolean`
    * required: `false`
    * default: `false`

    Says if the RTL/LTR should be reversed.

* **autoSize**
    
    * type: `Boolean`
    * required: `false`
    * default: `true`

    Says if the container height should swells and contracts to fit contents.

* **verticalCompact**
    
    * type: `Boolean`
    * required: `false`
    * default: `true`

    Says if the layout should be compact vertically.

* **preventCollision**
    
    * type: `Boolean`
    * required: `false`
    * default: `false`

    Says if grid items will move when being dragged over.

* **useCssTransforms**
    
    * type: `Boolean`
    * required: `false`
    * default: `true`

    Says if the CSS `transition-property: transform;` should be used.

* **responsive**
    
    * type: `Boolean`
    * required: `false`
    * default: `false`

    Says if the layout should be responsive to window width

* **breakpoints**

    * type: `Object`
    * required: `false`
    * default: { lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }

    Breakpoints defined for responsive layout. Sets widths on wich column number changes

* **cols**

    * type: `Object`
    * required: `false`
    * default: { lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }

    Defines number of columns for each breakpoint


##### GridItem

* **i**
    
    * type: `String`
    * required: `true`

    This is the unique identifier of the item.

* **x**
    
    * type: `Number`
    * required: `true`

    Says what is a initial horizontal position of the item (in which column it should be placed).

    The value must be a _whole number_. 

* **y**
    
    * type: `Number`
    * required: `true`

    Says what is a initial vertical position of the item (in which row it should be placed).

    The value must be a _whole number_. 

* **w**
    
    * type: `Number`
    * required: `true`

    Says what is a initial width of the item.

    The value is a number that is multiplied by `colWidth`.

* **h**
    
    * type: `Number`
    * required: `true`

    Says what is a initial height of the item.

    The value is a number that is multiplied by `rowHeight`.

* **minW**
    
    * type: `Number`
    * required: `false`
    * default: `1`

    Says what is a minimal width of the item. If `w` will be smaller then `minW` then `w` will be set to `minW`.

    The value is a number that is multiplied by `colWidth`.

* **minH**
    
    * type: `Number`
    * required: `false`
    * default: `1`

    Says what is a minimal hieght of the item. If `h` will be smaller then `minH` then `h` will be set to `minH`.

    The value is a number that is multiplied by `rowHeight`.

* **maxW**
    
    * type: `Number`
    * required: `false`
    * default: `Infinity`

    Says what is a maximal width of the item. If `w` will be bigger then `maxW` then `w` will be set to `maxW`.

    The value is a number that is multiplied by `colWidth`.

* **maxH**
    
    * type: `Number`
    * required: `false`
    * default: `Infinity`

    Says what is a maximal height of the item. If `h` will be bigger then `maxH` then `h` will be set to `maxH`.

    The value is a number that is multiplied by `rowHeight`

* **isDraggable**
    
    * type: `Boolean`
    * required: `false`
    * default: `null`

    Says if item is draggable.

    If default value is `null` then it's inherited from parent.

* **isResizable**
    
    * type: `Boolean`
    * required: `false`
    * default: `null`

    Says if item is resizable.

    If default value is `null` then it's inherited from parent.

* **static**
    
    * type: `Boolean`
    * required: `false`
    * default: `false`

    Says if item is static (won't be draggable, resizable or moved by other items).


* **dragIgnoreFrom**
    
    * type: `String`
    * required: `false`
    * default: `'a, button'`

    Says which elements of the item shouldn't trigger drag event of the item.

    The value is `css-like` selector string.

    For more info please refer to `ignoreFrom` in [interact.js docs](http://interactjs.io/docs/#ignorable-selectors).

* **dragAllowFrom**
    
    * type: `String`
    * required: `false`
    * default: `null`

    Says which elements of the item should trigger drag event of the item.

    The value is `css-like` selector string.
    
    If `null` then one can drag by any (excluding `dragIgnoreFrom`) element of the item.

    For more info please refer to `allowFrom` in [interact.js docs](http://interactjs.io/docs/#ignorable-selectors).

* **resizeIgnoreFrom**
    
    * type: `String`
    * required: `false`
    * default: `'a, button'`

    Says which elements of the item shouldn't trigger resize event of the item.

    The value is `css-like` selector string.

    For more info please refer to `ignoreFrom` in [interact.js docs](http://interactjs.io/docs/#ignorable-selectors).



#### Events

Move and resize event listeners can be added to each grid-item, so that the parent Vue can be notified when a grid element is being moved or resized.
Moved and resized event listeners can be added, if the only notification needed is when an item is finished moving or resizing.

Working example [here](https://jbaysolutions.github.io/vue-grid-layout/examples/02-events.html)   

````html

    <grid-layout
            :layout="layout"
            :col-num="12"
            :row-height="30"
            :is-draggable="true"
            :is-resizable="true"
            :vertical-compact="true"
            :margin="[10, 10]"
            :use-css-transforms="true"
            @layout-created="layoutCreatedEvent"
            @layout-before-mount="layoutBeforeMountEvent"
            @layout-mounted="layoutMountedEvent"
            @layout-ready="layoutReadyEvent"
            @layout-updated="layoutUpdatedEvent"
            @breakpoint-changed="breakpointChangedEvent"
    >

        <grid-item v-for="item in layout"
                   :x="item.x"
                   :y="item.y"
                   :w="item.w"
                   :h="item.h"
                   :i="item.i"
                   :key="item.i"
                   @resize="resizeEvent"
                   @move="moveEvent"
                   @resized="resizedEvent"
                   @container-resized="containerResizedEvent"
                   @moved="movedEvent">
            {{item.i}}
        </grid-item>
    </grid-layout>
```` 

* **layoutCreatedEvent**

    Layout created event

    Emited on the component created lifecycle hook

```javascript
    layoutCreatedEvent: function(newLayout){
      console.log("Created layout: ", newLayout)
    }
```

* **layoutBeforeMountEvent**

    Layout beforeMount event

    Emited on the component beforeMount lifecycle hook

```javascript
    layoutBeforeMountEvent: function(newLayout){
      console.log("beforeMount layout: ", newLayout)
    }
```

* **layoutMountedEvent**

    Layout mounted event

    Emited on the component mounted lifecycle hook

```javascript
    layoutMountedEvent: function(newLayout){
      console.log("Mounted layout: ", newLayout)
    }
```

* **layoutReadyEvent**

    Layout ready event

    Emited when all the operations on the mount hook finish

```javascript
    layoutReadyEvent: function(newLayout){
      console.log("Ready layout: ", newLayout)
    }
```

* **layoutUpdatedEvent**

    Layout updated event

    Every time the layout has finished updating and positions of all grid-items are recalculated

```javascript
    layoutUpdatedEvent: function(newLayout){
      console.log("Updated layout: ", newLayout)
    }
```

* **moveEvent**

    Move event

    Every time an item is being moved and changes position

```javascript
    moveEvent: function(i, newX, newY){
        console.log("MOVE i=" + i + ", X=" + newX + ", Y=" + newY);
    },
```

* **resizeEvent**

    Resize event

    Every time an item is being resized and changes size
 
```javascript
    resizeEvent: function(i, newH, newW, newHPx, newWPx){
        console.log("RESIZE i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
    },
``` 

* **movedEvent**

    Moved event

    Every time an item is finished being moved and changes position

```javascript
    movedEvent: function(i, newX, newY){
        console.log("MOVED i=" + i + ", X=" + newX + ", Y=" + newY);
    },
```

* **resizedEvent**

    Resized event

    Every time an item is finished being resized and changes size
 
```javascript
    /**
     * 
     * @param i the item id/index
     * @param newH new height in grid rows 
     * @param newW new width in grid columns
     * @param newHPx new height in pixels
     * @param newWPx new width in pixels
     * 
     */
    resizedEvent: function(i, newH, newW, newHPx, newWPx){
        console.log("RESIZED i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
    },
``` 

* **containerResizedEvent**

    Container Resized event

    Every time the grid item/layout container changes size (browser window or other)
 
```javascript
    /**
     * 
     * @param i the item id/index
     * @param newH new height in grid rows 
     * @param newW new width in grid columns
     * @param newHPx new height in pixels
     * @param newWPx new width in pixels
     * 
     */
    containerResizedEvent: function(i, newH, newW, newHPx, newWPx){
        console.log("CONTAINER RESIZED i=" + i + ", H=" + newH + ", W=" + newW + ", H(px)=" + newHPx + ", W(px)=" + newWPx);
    },
``` 

* **breakpointChangedEvent**

    Breakpoint Changed event

    Every time the breakpoint value changes due to window resize
 
```javascript
    /**
     * 
     * @param newBreakpoint the breakpoint name
     * @param newLayout the chosen layout for the breakpoint
     * 
     */
    breakpointChangedEvent: function(newBreakpoint, newLayout){
        console.log("BREAKPOINT CHANGED breakpoint=", newBreakpoint, ", layout: ", newLayout );
    },
``` 


## Contribute

If you have a feature request, please add it as an issue or make a pull request.


## TODO List

- [x] Basic grid layout
- [x] Responsive
- [x] Draggable grid items
- [x] Resizable grid items
- [x] Static elements
- [x] Min/max w/h per item


CHANGELOG.md
# Changelog

## 2.3.9 (September 28, 2020)

* Fix plugin installation (fixes #311) (thanks [yfwz100](https://github.com/yfwz100))
* Bumped interact.js to 1.9.22

## 2.3.8 (July 31, 2020)

* Make margins reactive (thanks [yfwz100](https://github.com/yfwz100))
* Added missing 'layout-updated' events (thanks [ben-lau](https://github.com/ben-lau))
* Support for initial responsive layouts and breakpoint change event (thanks [ftylitak](https://github.com/ftylitak))
* Bugfix for possible error when layout items get removed and interactObj is not set (thanks [Tofandel](https://github.com/Tofandel))
* Fix wrong grid item sorting in responsive mode (thanks [pieterbeulque](https://github.com/pieterbeulque))

## 2.3.7 (Oct 31, 2019)

* renamed 'resized' event emitted when the container changes size (browser window or other), introduced in #337, fixes #358 
* bugfix on layout-ready event

## 2.3.6 (Sep 11, 2019)

* reverted grid item w and h previous rounding change, fixes #355

## 2.3.5 (Sep 7, 2019)

* Fix: no-touch on mobile and only if item is draggable or resizable (thanks [DGoms](https://github.com/DGoms)).
* Prevent collision feature (thanks [SheanDe](https://github.com/SheanDe)).
* Simplified Chinese README (thanks [harrywangchina](https://github.com/harrywangchina)).
* 'resized' event now emitted if grid item is resized due to changes other than a manual resize (e.g. browser window resize) (thanks [pmorch](https://github.com/pmorch)).
* Improvement on 'layout-ready' event emit (thanks [pmorch](https://github.com/pmorch)).
* Added watchers for min/max height and width (thanks [grinat](https://github.com/grinat)).
* Improvement: make sure the size of grid-item is bigger than the slot-item (thanks [ywmail](https://github.com/ywmail)).

## 2.3.4 (Mar 5, 2019)

* Support for static items (thanks [panjiangyi](https://github.com/panjiangyi)).
* RTL bugfix (thanks [irvingwa](https://github.com/irvingwa)).
* Memory leak fixes (thanks [aiankile](https://github.com/aiankile)).
* Fixed exception on grid layout mount (thanks [BenoitZugmeyer](https://github.com/BenoitZugmeyer)).
* Fixed overlapping and resizing bugs on responsive mode (thanks [shpfive](https://github.com/shpfive)).
* Added new events emited by GridLayout (layout-created, layout-before-mount, layout-mounted, layout-ready) (thanks [samuelmolinski](https://github.com/samuelmolinski)).

## 2.3.3 (Dec 26, 2018)

* Reverted adding vue as external, which caused problems loading umd.

## 2.3.2 (Dec 13, 2018)

* Added vue as externals webpack config to fix startup problems (thanks [Micene9](https://github.com/Micene09))

## 2.3.1 (Nov 6, 2018)

* Previous fix for touch dragging on Android broke dragging on other mobile browsers (thanks [onx2](https://github.com/onx2))


## 2.3.0 (Oct 26, 2018)

* Responsive layout support (thanks [shpfive](https://github.com/shpfive))
* Fix for touch dragging on Android (thanks [pbabey](https://github.com/pbabey))

## 2.2.0 (Sep 10, 2018)

* changed project structure and build using Vue CLI
* GridItem new autosize method. For now, need to be called from child component with this.$parent.autoSize() (thanks[mech01nc01](https://github.com/mech01nc01)) 
* Abstract DOM related calls for initial SSR Support (thanks[Kukks](https://github.com/Kukks)) 
* GridItem.i can now be a number or a string (thanks[xch1029](https://github.com/xch1029)) 
* Use interactjs size restrictions to limit resizing (thanks[LuisCarreras](https://github.com/LuisCarreras)) 
* Fixed interactjs instance leak on instance release (thanks[zzuligy](https://github.com/zzuligy)) 

## 2.1.13 (May 2, 2018)

* Resize event now also emits dimensions in pixels (thanks [buremba](https://github.com/buremba)) 
* Support for dynamic col-num (#121) (thanks [ittus](https://github.com/ittus)) 

## 2.1.12 (Fev 27, 2018)

* Updated interact.js to 1.3.3 (#144) 

## 2.1.11 (Jan 5, 2018)

* Fixed issue with multiple grids on same vm (#134) (thanks [Suen](https://github.com/sunzongzheng))
* Fixed issue with layout update on reassignment (#130) (thanks [daizengyu](https://github.com/daizengyu123))

## 2.1.10 (Dec 15, 2017)

* Fixed possible bug related with #119
* Changed css translate to translate3d (#96)
* Added is-mirrored config, allowing the grid items to be rendered from right to left (horizontal flip) (thanks [kweij](https://github.com/kweij))
* Added grid updated event for easier integration with vuex (thanks [SergeyKhval](https://github.com/SergeyKhval))

## 2.1.9 (Aug 17, 2017)

* Fixed local module reference to interact.js

## 2.1.8 (Aug 17, 2017)

* Fixed #61 and #37
* Fixed #82
* Fixed #87

## 2.1.7 (Aug 17, 2017)

* Fixed #59
* Fixed #83
* Implemented support for dragAllowFrom and dragIgnoreFrom props on GridItem (thanks [ThePlastic](https://github.com/ThePlastic))

## 2.1.6 (Apr 6, 2017)

* Fixed #43, configurable drag elements ignore on grid item contents (thanks [neithere](https://github.com/neithere)) 
* Fix for getLayoutItem, sometimes returns null elements (thanks [pbabey](https://github.com/pbabey))

## 2.1.5 (Mar 24, 2017)

* Really fixed #22 #32, multiple grid instances were not working properly in 2.1.4
* resizedEvent now also returns item width and height in pixels (implements #34)


## 2.1.4 (Mar 20, 2017)

* Implemented #32, support for multiple grid instances on the same page

## 2.1.3 (Mar 9, 2017)

* Fixed #27, props mutation warnings


## 2.1.2 (Fev 16, 2017)

* Implemented #12, buttons on GridItems would trigger drag on mobile
* Implemented #24, listeners removal beforeDestroy (thanks [pbabey](https://github.com/pbabey))


## 2.1.1 (Fev 9, 2017)

* Implemented #13, dynamic row height update support
* Implemented #23, dynamic enable/disable dragging and resizing support
* Implemented #21, moved and resized events


## 2.1.0 (Fev 6, 2017)

* RTL support (thanks [easteregg](https://github.com/easteregg))
* Move and resize events (thanks [ThePlastic](https://github.com/ThePlastic))


OLD/src/app.js
var testLayout = [
    {"x":0,"y":0,"w":2,"h":2,"i":"0"},
    {"x":2,"y":0,"w":2,"h":4,"i":"1"},
    {"x":4,"y":0,"w":2,"h":5,"i":"2"},
    {"x":6,"y":0,"w":2,"h":3,"i":"3"},
    {"x":8,"y":0,"w":2,"h":3,"i":"4"},
    {"x":10,"y":0,"w":2,"h":3,"i":"5"},
    {"x":0,"y":5,"w":2,"h":5,"i":"6"},
    {"x":2,"y":5,"w":2,"h":5,"i":"7"},
    {"x":4,"y":5,"w":2,"h":5,"i":"8"},
    {"x":6,"y":4,"w":2,"h":4,"i":"9"},
    {"x":8,"y":4,"w":2,"h":4,"i":"10"},
    {"x":10,"y":4,"w":2,"h":4,"i":"11"},
    {"x":0,"y":10,"w":2,"h":5,"i":"12"},
    {"x":2,"y":10,"w":2,"h":5,"i":"13"},
    {"x":4,"y":8,"w":2,"h":4,"i":"14"},
    {"x":6,"y":8,"w":2,"h":4,"i":"15"},
    {"x":8,"y":10,"w":2,"h":5,"i":"16"},
    {"x":10,"y":4,"w":2,"h":2,"i":"17"},
    {"x":0,"y":9,"w":2,"h":3,"i":"18"},
    {"x":2,"y":6,"w":2,"h":2,"i":"19"}
];


//Vue.config.devtools = true;

import GridItem from './GridItem.vue';
import GridLayout from './GridLayout.vue';
//import ResponsiveGridLayout from './ResponsiveGridLayout.vue';
import TestElement from './TestElement.vue';
import CustomDragElement from './CustomDragElement.vue';

new Vue({
    el: '#app',
    components: {
        //ResponsiveGridLayout,
        GridLayout,
        GridItem,
        TestElement,
        CustomDragElement,
    },
    data: {
        layout: JSON.parse(JSON.stringify(testLayout)),
        colNum: 0,
        index: 0
    },
    mounted: function () {
        this.index = this.layout.length;
    },
    watch: {
    },
    methods: {
        increaseWidth: function(item) {
            var width = document.getElementById("content").offsetWidth;
            width += 20;
            document.getElementById("content").style.width = width+"px";
        },
        decreaseWidth: function(item) {

            var width = document.getElementById("content").offsetWidth;
            width -= 20;
            document.getElementById("content").style.width = width+"px";
        },
        removeItem: function(item) {
            //console.log("### REMOVE " + item.i);
            this.layout.splice(this.layout.indexOf(item), 1);
        },
        addItem: function() {
            var self = this;
            //console.log("### LENGTH: " + this.layout.length);
            var item = {"x":0,"y":0,"w":2,"h":2,"i":this.index+"", whatever: "bbb"};
            this.index++;
            this.layout.push(item);
        }
    },
});


/*function generateLayout() {
    return _.map(_.range(0, 25), function (item, i) {
        var y = Math.ceil(Math.random() * 4) + 1;
        return {
            x: _.random(0, 5) * 2 % 12,
            y: Math.floor(i / 6) * y,
            w: 2,
            h: y,
            i: i.toString(),
            static: Math.random() < 0.05
        };
    });
}*/



